### PR TITLE
Add functional swerve drive code

### PR DESCRIPTION
**Pull request by Gold856**

### CODE REVIEW

File: src/main/java/frc/robot/SwerveModule.java [hunk 1]
```diff
@@ -35,61 +39,93 @@ public SwerveModule(int CANport, int drivePort, int steerPort, double magnetOffs
...
-		state = SwerveModuleState.optimize(state, state.angle);
-		// Set drive speed
-		m_driveMotor.set(state.speedMetersPerSecond * DriveConstants.kDriveScale);
-		m_PIDController.setSetpoint(state.angle.getDegrees());
+		state = SwerveModuleState.optimize(state, Rotation2d.fromDegrees(getModuleAngle()));
+		// Set drive and steer speed
+		m_driveMotor.set(state.speedMetersPerSecond);
+		m_steerMotor.set(m_PIDController.calculate(getModuleAngle(), state.angle.getDegrees()));
 		// Print state to dashboard
 		SmartDashboard.putString("Swerve module " + m_CANCoder.getDeviceID(), state.toString());
```
Comment (JamisOrr) [inline src/main/java/frc/robot/SwerveModule.java]: Probably don't want to be printing here

File: src/main/java/frc/robot/commands/DriveDistanceCommand.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		}
+		var moduleStates = m_driveSubsystem.calculateModuleStates(new ChassisSpeeds(sign * speed, 0, 0), false);
+		m_driveSubsystem.setSwerveStates(moduleStates);
+	}
+
+	@Override
+	public boolean isFinished() {
+		// Determine whether the target distance has been reached
+		double diff = getDiff();
+		SmartDashboard.putNumber("diff", diff);
```
Comment (JamisOrr) [inline src/main/java/frc/robot/commands/DriveDistanceCommand.java]: Delete this

File: src/main/java/frc/robot/commands/DriveTimeCommand.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.subsystems.DriveSubsystem;
+
+public class DriveTimeCommand extends Command {
+	private DriveSubsystem m_driveSubsystem;
+	private double m_seconds;
+	private final Timer m_timer = new Timer();
+
+	/** Creates a new DriveTimeCommand. */
+	public DriveTimeCommand(DriveSubsystem subsystem, double seconds) {
```
Comment (JamisOrr) [inline src/main/java/frc/robot/commands/DriveTimeCommand.java]: Drive time should take in a time and a speed

Comment (JamisOrr) [reply src/main/java/frc/robot/commands/DriveTimeCommand.java]: I know the name is a little counter intuitive but its far more useful that way

**Review by JamisOrr — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -79,9 +80,9 @@ public static final class DriveConstants {
 		public static final double kDriveScale = 0.5;
 		// Drive PID values
 		public static final double kP = 0.005;
-		public static final double kI = 0;
+		public static final double kI = 0.045;
 		public static final double kD = 0;
-		public static final double kIz = 0;
+		public static final double kIz = 5;
```
Comment (JamisOrr) [inline src/main/java/frc/robot/Constants.java:85]: 5 meter IZone...

File: src/main/java/frc/robot/commands/DriveDistanceCommand.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
+package frc.robot.commands;
+
+import static frc.robot.Constants.DriveConstants.*;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.subsystems.DriveSubsystem;
+
+public class DriveDistanceCommand extends Command {
```
Comment (JamisOrr) [inline src/main/java/frc/robot/commands/DriveDistanceCommand.java:11]: Rename

File: src/main/java/frc/robot/commands/PIDTurnCommand.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.subsystems.DriveSubsystem;
+
+public class PIDTurnCommand extends Command {
+	private final double m_targetAngle;
+	private final double m_angleThreshold;
+	private final DriveSubsystem m_driveSubsystem;
+	private final PIDController m_contoller = new PIDController(0.01, 0, 0);
```
Comment (JamisOrr) [inline src/main/java/frc/robot/commands/PIDTurnCommand.java]: Test these constants?

File: src/main/java/frc/robot/commands/PIDTurnCommand.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	/**
+	 * Called every time the scheduler runs while the command is scheduled.
+	 */
+	@Override
+	public void execute() {
+		double rot = m_driveSubsystem.getHeading().getDegrees();
+		double speed = m_contoller.calculate(rot);
+
+		m_driveSubsystem.setModuleStates(0, 0, speed, false);
+		SmartDashboard.putNumber("Heading", rot);
```
Comment (JamisOrr) [inline src/main/java/frc/robot/commands/PIDTurnCommand.java:51]: We'll want to delete these once tuning is done

File: src/main/java/frc/robot/commands/TurnRelativeCommand.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	 * 
+	 * @param subsystem The drive subsystem.
+	 * @param angle     The angle in degrees.
+	 */
+	public TurnRelativeCommand(DriveSubsystem subsystem, double angle) {
+		m_driveSubsystem = subsystem;
+		m_angle = angle;
+		m_controller = new PIDController(0.0075, 0, 0);
+		m_controller.enableContinuousInput(0, 360);
+		m_controller.setTolerance(5);
```
Comment (JamisOrr) [inline src/main/java/frc/robot/commands/TurnRelativeCommand.java:28]: May want to lower this and keep tuning, absolute turn may be the better use of our time for right now though

File: src/main/java/frc/robot/subsystems/DriveSubsystem.java [hunk 9]
```diff
@@ -70,149 +70,232 @@ public DriveSubsystem() {
...
-				new SwerveModuleState(m_backLeft.getDriveSpeed(), Rotation2d.fromDegrees(m_backLeft.getModuleAngle())),
-				new SwerveModuleState(m_backRight.getDriveSpeed(),
-						Rotation2d.fromDegrees(m_backRight.getModuleAngle())) };
+		SmartDashboard.putNumber("Current Position", getModulePositions()[0].distanceMeters);
+		m_posePublisher.set(m_odometry.update(getHeading(), getModulePositions()));
+		SwerveModuleState[] states = { m_frontLeft.getModuleState(), m_frontRight.getModuleState(),
+				m_backLeft.getModuleState(), m_backRight.getModuleState() };
 		m_currentModuleStatePublisher.set(states);
+		SmartDashboard.putNumber("Steer 1 motor current", m_frontLeft.getSteerCurrent());
+		SmartDashboard.putNumber("Steer 3 motor current", m_frontRight.getSteerCurrent());
```
Comment (JamisOrr) [inline src/main/java/frc/robot/subsystems/DriveSubsystem.java:244]: Delete these now that debugging the turning is done

File: src/test/java/SwerveTest.java [hunk 10]
```diff
@@ -2,38 +2,67 @@
...
 import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.simulation.DriverStationSim;
+import edu.wpi.first.wpilibj.simulation.SimHooks;
+import edu.wpi.first.wpilibj2.command.CommandScheduler;
+import frc.robot.commands.DriveDistanceCommand;
+import frc.robot.commands.DriveTimeCommand;
 import frc.robot.subsystems.DriveSubsystem;
 
 public class SwerveTest {
```
Comment (JamisOrr) [inline src/test/java/SwerveTest.java:20]: Is this whole file just to save you some time when you were doing repeated testing of the swerve modules? Should we keep this in when we merge into main?

Comment (Gold856) [reply src/test/java/SwerveTest.java:20]: Unit tests are important. This will be expanded later on.

**Pull request closed without merging**